### PR TITLE
Added impression counter

### PR DESCRIPTION
**Pull request by dortheimer**
Adding impression counter and view erosion for interestingness measurements.

### CODE REVIEW

File: api/model/impression.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+            ip: ["required", "string"],
+
+        };
+    }
+
+    get tableName() {
+        return "impression";
+    }
+
+    person() {
```
Comment (florpor) [inline api/model/impression.js:19]: impression doesn't necessarily link to valid person. suggest this be removed

File: api/controller/impression.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    /**
+     * Aggregate impressions and update view count in plan table.
+     * Should be executed by cron every hour or day
+     */
+    aggregate() {
+        
+        // get aggregated count for every plan
+        const query = 'select a.plan_id, count(*) as num from (SELECT plan_id, ip FROM `impression` group by plan_id, ip) as a group by a.plan_id ';
```
Comment (florpor) [inline api/controller/impression.js:31]: can avoid inner query with:
`SELECT plan_id, COUNT(DISTINCT ip) FROM impression`

File: api/controller/impression.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
+const Bluebird = require("bluebird");
+const Controller = require("../controller/controller");
+const Impression = require("../model/impression");
+const Plan = require("../model/plan");
+const { Knex } = require('../service/database')
+
+class ImpressionController extends Controller {
+
+    getIp = (req) => {
+        return req.headers["x-forwarded-for"] || req.connection.remoteAddress;
```
Comment (florpor) [inline api/controller/impression.js:10]: we wanted to hash the ip, no?
we don't want to keep actual client ips, plus it can save space

File: api/controller/impression.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                        views=views+${key}, 
+                        erosion_views=erosion_views +${key} 
+                        WHERE id IN(${map[key].join(',')})`);
+
+                // update the plan table
+                return Bluebird.mapSeries(queries, query => Knex.raw(query))
+            })
+            .then(() => {
+                // remove impressions
+                return Knex.raw('TRUNCATE impression')
```
Comment (florpor) [inline api/controller/impression.js:57]: new impressions created while update plan commands are running (can be hundreds of them) will be deleted and never counted.
do you think this is reasonable, or should a harsher mechanism be implemented?

File: api/model/plan.js [hunk 5]
```diff
@@ -16,7 +18,9 @@ class Plan extends Model {
             geom: ["required", "object"],
             jurisdiction: "string",
             areaChanges: "string",
-            rating: ["required", "number"]
+            rating: ["required", "number"],
+            views: ["required", "number"],
+            erosion_views: ["required", "number"]
```
Comment (florpor) [inline api/model/plan.js:23]: what is the meaning of this field?

File: src/scenes/SinglePlan.jsx [hunk 6]
```diff
@@ -99,6 +102,7 @@ class SinglePlan extends Component {
 
         changes &&
             changes[0].forEach(change => {
+                if (!change[3]) return;
```
Comment (florpor) [inline src/scenes/SinglePlan.jsx:105]: this shouldn't be part of this pr

**Review by florpor — CHANGES_REQUESTED** [review]

File: api/controller/impression.js (hunk 4 above)
Comment (dortheimer) [reply api/controller/impression.js]: Added a transaction. But it think it's reasonable.

File: src/scenes/SinglePlan.jsx (hunk 6 above)
Comment (dortheimer) [reply src/scenes/SinglePlan.jsx:110]: I bumped into a bug...

Comment (florpor) [reply src/scenes/SinglePlan.jsx:110]: cool. the added comment is good

**Review by florpor — APPROVED** [review]
